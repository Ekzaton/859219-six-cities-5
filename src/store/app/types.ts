import {ActionType} from "../const";
import {FilteringType, SortingType} from "../../const";

export type AppState = {
  activeOfferID: number | null,
  currentFiltering: FilteringType,
  currentSorting: SortingType
};

type getActiveOfferIDAction = {
  type: ActionType.GET_ACTIVE_OFFER_ID,
  payload: number | null
}

type getCurrentFilteringAction = {
  type: ActionType.GET_CURRENT_FILTERING,
  payload: FilteringType
}

type getCurrentSortingAction = {
  type: ActionType.GET_CURRENT_SORTING,
  payload: SortingType
}

export type AppAction = getActiveOfferIDAction | getCurrentFilteringAction | getCurrentSortingAction;
