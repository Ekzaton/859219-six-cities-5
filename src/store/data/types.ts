import {ActionType} from "../const";
import {Offer} from "../../components/types";

export type DataState = {
  offers: Offer[]
};

type getOffersAction = {
  type: typeof ActionType.GET_OFFERS,
  payload: Offer[]
}

export type DataAction = getOffersAction;

