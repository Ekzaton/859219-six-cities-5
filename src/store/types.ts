import {ActionType} from "./const";
import {Offer} from "../components/types";

export type State = {
  currentFilter: string,
  currentSorting: string,
  offers: Offer[],
};

type getCurrentFilterAction = {
  type: typeof ActionType.GET_CURRENT_FILTER,
  payload: string
}

type getCurrentSorting = {
  type: typeof ActionType.GET_CURRENT_SORTING,
  payload: string
}

type getOffersAction = {
  type: typeof ActionType.GET_OFFERS,
  payload: Offer[]
}

export type Action = getCurrentFilterAction | getCurrentSorting | getOffersAction;
