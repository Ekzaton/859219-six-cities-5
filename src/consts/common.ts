export enum AppRoute {
  MAIN = `/`,
  LOGIN = `/login/`,
  FAVORITES = `/favorites/`,
  OFFER_ID = `/offer/:id/`,
  OFFER = `/offer/`
}

export enum AuthStatus {
  AUTH = `auth`,
  NO_AUTH = `no-auth`
}

export enum FilteringType {
  PARIS = `Paris`,
  COLOGNE = `Cologne`,
  BRUSSELS = `Brussels`,
  AMSTERDAM = `Amsterdam`,
  HAMBURG = `Hamburg`,
  DUSSELDORF = `Dusseldorf`
}

export enum SortingType {
  POPULAR = `Popular`,
  PRICE_LOW_TO_HIGH = `Price: low to high`,
  PRICE_HIGH_TO_LOW = `Price: high to low`,
  TOP_RATED_FIRST = `Top rated first`
}
