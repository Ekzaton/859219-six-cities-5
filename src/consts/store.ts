export enum ActionType {
  CHANGE_ACTIVE_OFFER_ID = `CHANGE_ACTIVE_OFFER_ID`,
  CHANGE_CURRENT_CITY = `CHANGE_CURRENT_CITY`,
  CHANGE_CURRENT_SORTING = `CHANGE_CURRENT_SORTING`,
  CHANGE_OFFER_FAVORITE_STATUS = `CHANGE_OFFER_FAVORITE_STATUS`,
  GET_ALL_OFFERS = `GET_ALL_OFFERS`,
  GET_FAVORITE_OFFERS = `GET_FAVORITE_OFFERS`,
  GET_NEARBY_OFFERS = `GET_NEARBY_OFFERS`,
  GET_SINGLE_OFFER = `GET_SINGLE_OFFER`,
  GET_REVIEWS = `GET_REVIEWS`,
  GET_USER = `GET_USER`,
  REDIRECT_TO_ROUTE = `REDIRECT_TO_ROUTE`,
  SET_AUTH_STATUS = `SET_AUTH_STATUS`,
  SET_INITIAL_STATE = `SET_INITIAL_STATE`,
  SET_IS_ALL_OFFERS_LOADING = `SET_IS_ALL_OFFERS_LOADING`,
  SET_IS_ALL_OFFERS_LOADING_ERROR = `SET_IS_ALL_OFFERS_LOADING_ERROR`,
  SET_IS_FAVORITE_OFFERS_LOADING = `SET_IS_FAVORITE_OFFERS_LOADING`,
  SET_IS_FAVORITE_OFFERS_LOADING_ERROR = `SET_IS_FAVORITE_OFFERS_LOADING_ERROR`,
  SET_IS_FAVORITE_SENDING = `SET_IS_FAVORITE_SENDING`,
  SET_IS_FAVORITE_SENDING_ERROR = `SET_IS_FAVORITE_SENDING_ERROR`,
  SET_IS_LOGIN_SENDING = `SET_IS_LOGIN_SENDING`,
  SET_IS_LOGIN_SENDING_ERROR = `SET_IS_LOGIN_SENDING_ERROR`,
  SET_IS_REVIEW_SENDING = `SET_IS_REVIEW_SENDING`,
  SET_IS_REVIEW_SENDING_ERROR = `SET_IS_REVIEW_SENDING_ERROR`,
  SET_IS_SINGLE_OFFER_LOADING = `SET_IS_SINGLE_OFFER_LOADING`,
  SET_IS_SINGLE_OFFER_LOADING_ERROR = `SET_IS_SINGLE_OFFER_LOADING_ERROR`
}

export enum APIEndpoint {
  HOTELS = `/hotels/`,
  NEARBY = `/nearby/`,
  FAVORITE = `/favorite/`,
  COMMENTS = `/comments/`,
  LOGIN = `/login/`,
  LOGOUT = `/logout/`
}
