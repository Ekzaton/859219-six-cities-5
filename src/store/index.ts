import {AxiosInstance} from "axios";
import {applyMiddleware, combineReducers, createStore} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import thunk, {ThunkAction, ThunkMiddleware} from "redux-thunk";

import {appReducer} from "./app/reducer";
import {dataReducer} from "./data/reducer";
import {userReducer} from "./user/reducer";

import {AppAction} from "./app/types";
import {DataAction} from "./data/types";
import {UserAction} from "./user/types";

import {redirect} from "../middlewares/redirect";
import {createAPI} from "../services/api";

const api = createAPI();
const thunkWithAPI = thunk.withExtraArgument(api);
const rootReducer = combineReducers({app: appReducer, data: dataReducer, user: userReducer});
const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunkWithAPI as APIMiddleware, redirect)));

export type RootState = ReturnType<typeof rootReducer>;
type Action = AppAction | DataAction | UserAction;
type APIMiddleware = ThunkMiddleware<RootState, Action, AxiosInstance>;
export type APIAction = ThunkAction<void, RootState, AxiosInstance, Action>;

export default store;
