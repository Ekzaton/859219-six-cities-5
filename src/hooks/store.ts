import {useDispatch, useSelector, TypedUseSelectorHook} from "react-redux";

import {AppDispatch, AppState} from "../store/store";

type TypedUseDispatchHook<T> = () => (...args: unknown[]) => T;

export const useAppDispatch: TypedUseDispatchHook<AppDispatch> = useDispatch;
export const useAppSelector: TypedUseSelectorHook<AppState> = useSelector;
