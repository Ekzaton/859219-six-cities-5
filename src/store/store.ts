import {AxiosInstance} from "axios";
import {applyMiddleware, combineReducers, createStore} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import thunk, {ThunkAction} from "redux-thunk";

import {redirect} from "../middlewares/redirect";
import {createAPI} from "../services/api";
import {FavoritesAction} from "../types/store/favorites";
import {LoginAction} from "../types/store/login";
import {MainAction} from "../types/store/main";
import {PropertyAction} from "../types/store/property";

import {favoritesReducer} from "./favorites/reducer";
import {loginReducer} from "./login/reducer";
import {mainReducer} from "./main/reducer";
import {propertyReducer} from "./property/reducer";

const api = createAPI();
const thunkWithAPI = thunk.withExtraArgument(api);

const rootReducer = combineReducers({
  favorites: favoritesReducer,
  login: loginReducer,
  main: mainReducer,
  property: propertyReducer,
});

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunkWithAPI, redirect)));

export type AppState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppAction = FavoritesAction | LoginAction | MainAction | PropertyAction;
export type AppThunkAction = ThunkAction<void, AppState, AxiosInstance, AppAction>;
export default store;
