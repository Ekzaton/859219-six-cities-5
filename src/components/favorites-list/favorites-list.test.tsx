import React from 'react';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';
import createMockStore from 'redux-mock-store';
import {render} from '@testing-library/react';

import {AuthStatus} from "../../consts/common";
import offersMock from "../../mock/offers.json";
import {Offer} from "../../types/common";
import {getOffersByCity} from "../../utils/store";

import FavoritesList from "./favorites-list";

describe(`Favorites list renders correctly`, () => {
  const mockStore = createMockStore([]);
  const offersByCity = getOffersByCity(offersMock as Offer[]);

  it(`for authorized user`, () => {
    const store = mockStore({
      login: {
        authStatus: AuthStatus.AUTH
      }
    });

    const {container} = render(
        <Provider store={store}>
          <MemoryRouter>
            <FavoritesList
              offersByCity={offersByCity}
            />
          </MemoryRouter>
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
