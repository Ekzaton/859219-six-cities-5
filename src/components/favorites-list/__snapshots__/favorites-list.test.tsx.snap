// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites list renders correctly for authorized user 1`] = `
<div>
  <ul
    class="favorites__list"
  >
    <li
      class="favorites__locations-items"
    >
      <div
        class="favorites__locations locations locations--current"
      >
        <div
          class="locations__item"
        >
          <a
            class="locations__item-link"
            href="/"
          >
            <span>
              Amsterdam
            </span>
          </a>
        </div>
      </div>
      <div
        class="favorites__places"
      >
        <article
          class="favorites__card place-card"
        >
          <div
            class="favorites__image-wrapper place-card__image-wrapper"
          >
            <a
              href="/offer/1"
            >
              <img
                alt="Place image"
                class="place-card__image"
                height="110"
                src="img/apartment-01.jpg"
                width="150"
              />
            </a>
          </div>
          <div
            class="favorites_card-info place-card__info"
          >
            <div
              class="place-card__price-wrapper"
            >
              <div
                class="place-card__price"
              >
                <b
                  class="place-card__price-value"
                >
                  €
                  120
                </b>
                <span
                  class="place-card__price-text"
                >
                   / night
                </span>
              </div>
              <button
                class="place-card__bookmark-button false button"
                type="button"
              >
                <svg
                  class="place-card__bookmark-icon"
                  height="19"
                  width="18"
                >
                  <use
                    xlink:href="#icon-bookmark"
                  />
                </svg>
                <span
                  class="visually-hidden"
                >
                  To bookmarks
                </span>
              </button>
            </div>
            <div
              class="place-card__rating rating"
            >
              <div
                class="place-card__stars rating__stars"
              >
                <span
                  style="width: 96%;"
                />
                <span
                  class="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <h2
              class="place-card__name"
            >
              <a
                href="/offer/1"
              >
                Beautiful & luxurious studio at great location
              </a>
            </h2>
            <p
              class="place-card__type"
            >
              Apartment
            </p>
          </div>
        </article>
        <article
          class="favorites__card place-card"
        >
          <div
            class="place-card__mark"
          >
            <span>
              Premium
            </span>
          </div>
          <div
            class="favorites__image-wrapper place-card__image-wrapper"
          >
            <a
              href="/offer/2"
            >
              <img
                alt="Place image"
                class="place-card__image"
                height="110"
                src="img/apartment-02.jpg"
                width="150"
              />
            </a>
          </div>
          <div
            class="favorites_card-info place-card__info"
          >
            <div
              class="place-card__price-wrapper"
            >
              <div
                class="place-card__price"
              >
                <b
                  class="place-card__price-value"
                >
                  €
                  180
                </b>
                <span
                  class="place-card__price-text"
                >
                   / night
                </span>
              </div>
              <button
                class="place-card__bookmark-button false button"
                type="button"
              >
                <svg
                  class="place-card__bookmark-icon"
                  height="19"
                  width="18"
                >
                  <use
                    xlink:href="#icon-bookmark"
                  />
                </svg>
                <span
                  class="visually-hidden"
                >
                  To bookmarks
                </span>
              </button>
            </div>
            <div
              class="place-card__rating rating"
            >
              <div
                class="place-card__stars rating__stars"
              >
                <span
                  style="width: 100%;"
                />
                <span
                  class="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <h2
              class="place-card__name"
            >
              <a
                href="/offer/2"
              >
                Wood and stone place
              </a>
            </h2>
            <p
              class="place-card__type"
            >
              Hotel
            </p>
          </div>
        </article>
        <article
          class="favorites__card place-card"
        >
          <div
            class="favorites__image-wrapper place-card__image-wrapper"
          >
            <a
              href="/offer/3"
            >
              <img
                alt="Place image"
                class="place-card__image"
                height="110"
                src="img/apartment-03.jpg"
                width="150"
              />
            </a>
          </div>
          <div
            class="favorites_card-info place-card__info"
          >
            <div
              class="place-card__price-wrapper"
            >
              <div
                class="place-card__price"
              >
                <b
                  class="place-card__price-value"
                >
                  €
                  100
                </b>
                <span
                  class="place-card__price-text"
                >
                   / night
                </span>
              </div>
              <button
                class="place-card__bookmark-button false button"
                type="button"
              >
                <svg
                  class="place-card__bookmark-icon"
                  height="19"
                  width="18"
                >
                  <use
                    xlink:href="#icon-bookmark"
                  />
                </svg>
                <span
                  class="visually-hidden"
                >
                  To bookmarks
                </span>
              </button>
            </div>
            <div
              class="place-card__rating rating"
            >
              <div
                class="place-card__stars rating__stars"
              >
                <span
                  style="width: 80%;"
                />
                <span
                  class="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <h2
              class="place-card__name"
            >
              <a
                href="/offer/3"
              >
                Canal View Prinsengracht
              </a>
            </h2>
            <p
              class="place-card__type"
            >
              House
            </p>
          </div>
        </article>
        <article
          class="favorites__card place-card"
        >
          <div
            class="favorites__image-wrapper place-card__image-wrapper"
          >
            <a
              href="/offer/4"
            >
              <img
                alt="Place image"
                class="place-card__image"
                height="110"
                src="img/room.jpg"
                width="150"
              />
            </a>
          </div>
          <div
            class="favorites_card-info place-card__info"
          >
            <div
              class="place-card__price-wrapper"
            >
              <div
                class="place-card__price"
              >
                <b
                  class="place-card__price-value"
                >
                  €
                  70
                </b>
                <span
                  class="place-card__price-text"
                >
                   / night
                </span>
              </div>
              <button
                class="place-card__bookmark-button place-card__bookmark-button--active button"
                type="button"
              >
                <svg
                  class="place-card__bookmark-icon"
                  height="19"
                  width="18"
                >
                  <use
                    xlink:href="#icon-bookmark"
                  />
                </svg>
                <span
                  class="visually-hidden"
                >
                  In bookmarks
                </span>
              </button>
            </div>
            <div
              class="place-card__rating rating"
            >
              <div
                class="place-card__stars rating__stars"
              >
                <span
                  style="width: 84%;"
                />
                <span
                  class="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <h2
              class="place-card__name"
            >
              <a
                href="/offer/4"
              >
                Nice, cozy, warm big bed apartment
              </a>
            </h2>
            <p
              class="place-card__type"
            >
              Room
            </p>
          </div>
        </article>
      </div>
    </li>
  </ul>
</div>
`;
